src/base/lane.cc

```diff
@@ -44,8 +44,28 @@ double Lane::do_length() const { return lane_geometry_->ArcLength(); }
 maliput::api::RBounds Lane::do_lane_bounds(double s) const { return lane_geometry_->RBounds(s); }
 
 maliput::api::RBounds Lane::do_segment_bounds(double s) const {
-  // TODO: Implement
-  MALIPUT_THROW_MESSAGE("Not implemented");
+  const maliput::api::RBounds lane_bounds = do_lane_bounds(s);
+  double bound_left = lane_bounds.max();
+  maliput::api::Lane const* other_lane = const_cast<maliput::api::Lane const*>(to_left());
+  while (other_lane != nullptr) {
+    const maliput::api::RBounds other_lane_bounds = other_lane->lane_bounds(s);
+    bound_left += other_lane_bounds.max() - other_lane_bounds.min();
+    other_lane = const_cast<maliput::api::Lane const*>(to_left());
+  }
+
+  double bound_right = -lane_bounds.min();
+  other_lane = const_cast<maliput::api::Lane const*>(to_right());
+  while (other_lane != nullptr) {
+    const maliput::api::RBounds other_lane_bounds = other_lane->lane_bounds(s);
+    bound_right += other_lane_bounds.max() - other_lane_bounds.min();
+    other_lane = const_cast<maliput::api::Lane const*>(to_right());
+  }
+
+  const double tolerance = lane_geometry_->linear_tolerance();
+  bound_left = bound_left < tolerance ? tolerance : bound_left;
+  bound_right = bound_right < tolerance ? tolerance : bound_right;
+
+  return {-bound_right, bound_left};
```

Comment:
The problem with this implementation is that we don't have any guarantee at construct time of the adjacency of Lanes.
We could consider changing the Builder to do so, or to force after the fact via a check to adjancent LaneGeometries to have a common LineString. That said, it is expected that Lanes are adjacent without any gaps.

I wonder if we should consider adding a function that implements the max - min because this is repeated in many places.

I wonder if we should add an interator interface to do things like this composing with the stl:

```C++
const maliput::api::RBounds lane_bounds = do_lane_bounds(s);
auto lane_it = this->segment()->lane_iterator(do_index());
auto add_lane_bounds = [s](double bound, auto lane_it) { return std:move(bound) + lane->lane_bounds(s).distance(); };
const double bound_left = std::accumulate(lane_it, segment()->lane_end_iterator(),  lane_bounds.max(), add_lane_bounds);
const double bound_right = std::accumulate(segment()->lane_start_iterator(),  std::prev(lane_it), -lane_bounds.min(), add_lane_bounds);
return {-bound_right, bound_left};
```
There are other places in which we do for each and we don't benefit form the STL algorithms because of not having a proper iterator interface. We should probably discuss it.

Reply:
> The problem with this implementation is that we don't have any guarantee at construct time of the adjacency of Lanes.

That's correct. However, for computing the segment bounds the maliput api is being used, and for such, it is expected that no gaps are present between lanes in a segment. For so I think that the builder is the right place to do this assertion 

> I wonder if we should consider adding a function that implements the max - min because this is repeated in many places.

We could provide a generic method for computing this however I think that this will end up relying on the backend implementation. In [maliput_malidrive](https://github.com/maliput/maliput_malidrive/blob/814882a828feded7949e51d7eeef6414a2b701f5/src/maliput_malidrive/base/lane.cc#L141) we rely on the conversion between the reference curve and the `s` in the correspondent lane.

> I wonder if we should add an interator interface

Agree, if we have a collection under the hood that at some point will be iterated, then having iterators could be convenient. 
However, while I was writing the issue in `maliput` for adding this I retracted it because, from a user point of view I couldn't think of a case where the lanes will be wanted to be iterated. It definitely doesn't harm the API but one use case could support the idea of this addition.

Reply:
For the implementation I pushed I have to make some corrections:
```cpp
const maliput::api::RBounds other_lane_bounds = other_lane->lane_bounds(s);
```
This isn't accurate, I shouldn't use the same `s` coordinate for all the lanes in the segment. One alternative is to scale the s value according to the arclength of the lane.

Reply:
> This isn't accurate, I shouldn't use the same s coordinate for all the lanes in the segment. One alternative is to scale the s value according to the arclength of the lane.

I missed it, good that you did. Correct, we need to scale it.